File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/FollowersListViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.model.UserInfo
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestoreUserRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class FollowersListType {
    FOLLOWERS,
    FOLLOWING
}

data class FollowersListUiState(
    val users: List<UserInfo> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val error: String? = null,
    val hasMore: Boolean = true,
    val currentPage: Int = 0,
    val isRefreshing: Boolean = false
)

class FollowersListViewModel(
    application: Application,
    private val userId: String,
    private val type: FollowersListType
) : AndroidViewModel(application) {

    private val prefs = AndroidPreferencesRepository(application.applicationContext)
    private val authRepository = AndroidAuthRepository(
        context = application.applicationContext,
        activity = null,
        preferencesRepository = prefs
    )
    private val userRepository = FirestoreUserRepository(authRepository = authRepository)

    private val _uiState = MutableStateFlow(FollowersListUiState())
    val uiState: StateFlow<FollowersListUiState> = _uiState.asStateFlow()

    private val pageSize = 20

    init {
        loadUsers()
    }

    fun loadUsers() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null, currentPage = 0)
            try {
                val users = when (type) {
                    FollowersListType.FOLLOWERS -> userRepository.getFollowers(userId, page = 0, pageSize = pageSize)
                    FollowersListType.FOLLOWING -> userRepository.getFollowing(userId, page = 0, pageSize = pageSize)
                }
                _uiState.value = _uiState.value.copy(
                    users = users,
                    isLoading = false,
                    hasMore = users.size >= pageSize,
                    currentPage = 0
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load ${type.name.lowercase()}"
                )
            }
        }
    }

    fun loadMore() {
        if (_uiState.value.isLoadingMore || !_uiState.value.hasMore) return

        viewModelScope.launch {
            val nextPage = _uiState.value.currentPage + 1
            _uiState.value = _uiState.value.copy(isLoadingMore = true, error = null)
            try {
                val newUsers = when (type) {
                    FollowersListType.FOLLOWERS -> userRepository.getFollowers(userId, page = nextPage, pageSize = pageSize)
                    FollowersListType.FOLLOWING -> userRepository.getFollowing(userId, page = nextPage, pageSize = pageSize)
                }
                _uiState.value = _uiState.value.copy(
                    users = _uiState.value.users + newUsers,
                    isLoadingMore = false,
                    hasMore = newUsers.size >= pageSize,
                    currentPage = nextPage
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoadingMore = false,
                    error = e.message ?: "Failed to load more ${type.name.lowercase()}"
                )
            }
        }
    }

    fun refresh() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isRefreshing = true, error = null, currentPage = 0)
            try {
                val users = when (type) {
                    FollowersListType.FOLLOWERS -> userRepository.getFollowers(userId, page = 0, pageSize = pageSize)
                    FollowersListType.FOLLOWING -> userRepository.getFollowing(userId, page = 0, pageSize = pageSize)
                }
                _uiState.value = _uiState.value.copy(
                    users = users,
                    isRefreshing = false,
                    hasMore = users.size >= pageSize,
                    currentPage = 0
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isRefreshing = false,
                    error = e.message ?: "Failed to refresh ${type.name.lowercase()}"
                )
            }
        }
    }
}

File: iosApp/iosApp/viewmodel/FollowersListViewModel.swift
import Foundation
import SwiftUI
import Shared

enum FollowersListType {
    case followers
    case following
}

@MainActor
class FollowersListViewModel: ObservableObject {
    @Published var users: [UserInfo] = []
    @Published var isLoading: Bool = false
    @Published var isLoadingMore: Bool = false
    @Published var error: String?
    @Published var hasMore: Bool = true
    @Published var isRefreshing: Bool = false
    
    private let userId: String
    private let type: FollowersListType
    private let pageSize: Int32 = 20
    private var currentPage: Int32 = 0
    
    private let prefs = IOSPreferencesRepository()
    private lazy var authRepo = IOSAuthRepository(preferencesRepository: prefs)
    private lazy var userRepo = FirestoreUserRepository(authRepository: authRepo)
    
    init(userId: String, type: FollowersListType) {
        self.userId = userId
        self.type = type
        Task {
            await loadUsers()
        }
    }
    
    func loadUsers() async {
        isLoading = true
        error = nil
        currentPage = 0
        
        do {
            let fetchedUsers = try await fetchUsers(page: 0)
            users = fetchedUsers
            isLoading = false
            hasMore = fetchedUsers.count >= pageSize
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }
    
    func loadMore() async {
        guard !isLoadingMore, hasMore else { return }
        
        let nextPage = currentPage + 1
        isLoadingMore = true
        error = nil
        
        do {
            let newUsers = try await fetchUsers(page: nextPage)
            users.append(contentsOf: newUsers)
            isLoadingMore = false
            hasMore = newUsers.count >= pageSize
            currentPage = nextPage
        } catch {
            self.error = error.localizedDescription
            isLoadingMore = false
        }
    }
    
    func refresh() async {
        isRefreshing = true
        error = nil
        currentPage = 0
        
        do {
            let fetchedUsers = try await fetchUsers(page: 0)
            users = fetchedUsers
            isRefreshing = false
            hasMore = fetchedUsers.count >= pageSize
        } catch {
            self.error = error.localizedDescription
            isRefreshing = false
        }
    }
    
    private func fetchUsers(page: Int32) async throws -> [UserInfo] {
        switch type {
        case .followers:
            return try await userRepo.getFollowers(userId: userId, page: page, pageSize: pageSize)
        case .following:
            return try await userRepo.getFollowing(userId: userId, page: page, pageSize: pageSize)
        }
    }
}
